### PR TITLE
 While-lists Python Libraries

### LINKED ISSUE #2832: Add 'copulas' library
#### Expected Behavior
The follwoing packages are whitelisted in QuantConnect:
- [copulas](https://pypi.org/project/copulas/) ([docs](https://dai-lab.github.io/Copulas/))
- [hmm](https://github.com/hmmlearn/hmmlearn)
- [pomegranate](https://pomegranate.readthedocs.io/en/latest/index.html)

#### Actual Behavior
Not available.

#### Potential Solution
Adds the following in [DockerfileLeanFoundation](https://github.com/QuantConnect/Lean/blob/master/DockerfileLeanFoundation#L79):
```
pip install copulas
pip install --upgrade --user hmmlearn
conda install pomegranate
```
Testing the potential solution:
- Build the docker image with the dockerfile after the changes.

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### LINKED ISSUE #3084: Add `tsfresh` to Foundation
Review package tsfresh for inclusion in the foundation.

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### LINKED ISSUE #3173: Add `LightGBM` Python Library to Foundation
#### Requested Behavior

Support: LightGBM
https://github.com/Microsoft/LightGBM
Python API: https://lightgbm.readthedocs.io/en/latest/index.html
Can be installed from PyPI - pip install lightgbm [ https://pypi.org/project/lightgbm/ ]

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### LINKED ISSUE #3178: Add BeautifulSoup Python Library to Foundation
#### Requested Behavior

Support:
[Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/) is a Python library for pulling data out of HTML and XML files. It works with your favorite parser to provide idiomatic ways of navigating, searching and modifying the parse tree. It commonly saves programmers hours or days of work.

Can be installed from PyPI - [pip install beautifulsoup4](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#installing-beautiful-soup)

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### LINKED ISSUE #3199: Add Facebook/Prophet Python Library to Foundation
Add Python library Prophet 
https://github.com/facebook/prophet
Can be installed via PyPI
`pip install fbprophet`

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
- [x] I have provided detailed steps to reproduce the issue

### LINKED ISSUE #3233: Add fastai Library
#### Expected Behavior

Add fastai Python library to lean. Powerful AI library available via
```
pip3 install https://download.pytorch.org/whl/cu90/torch-1.1.0-cp36-cp36m-win_amd64.whl
pip install fastai
```
Potential problem:
 -Requires matplotlib 2.2.2 and pandas 0.23.0, which is incompatible with alphastreams 0.8.0 requirements.
#### Checklist

- [x] I have completely filled out this template
- [ ] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [ ] I have provided detailed steps to reproduce the issue

### LINKED ISSUE #3234: Add pyramid-arima Library
#### Expected Behavior

Adds pyramid-arime library to supported Python libraries. Library's main functionality is to bring R's _auto.arima_ function into Python.
Available via PyPi
```
pip install pyramid-arima
```
#### Checklist

- [x] I have completely filled out this template
- [ ] I have confirmed that this issue exists on the current `master` branch
- [ ] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [ ] I have provided detailed steps to reproduce the issue

### LINKED ISSUE #3367: Add dependencies for Library NLTK
#### Expected Behavior

API methods of NLTK need dependencies (listed below). This can be done by commands:
```
import nltk
nltk.download('all')
```

The details of dependencies:

```punkt
averaged_perceptron_tagger
averaged_perceptron_tagger_ru
maxent_treebank_pos_tagger
universal_tagset
porter_test
rslp
vader_lexicon
bllip_wsj_no_aux
moses_sample
wmt15_eval
word2vec_sample
mwa_ppdb
perluniprops
tarsets
basque_grammars
book_grammars
large_grammars
sample_grammars
spanish_grammars
maxent_ne_chunker
abc
alpino
biocreative_ppi
brown
brown_tei
cess_cat
cess_esp
chat80
city_database
cmudict
comparative_sentences
conll2000
conll2002
crubadan
dependency_treebank
dolch
europarl_raw
floresta
framenet_v15
framenet_v17
gazetteers
genesis
gutenberg
ieer
inaugural
indian
kimmo
lin_thesaurus
mac_morpho
movie_reviews
mte_teip5
names
nonbreaking_prefixes
nps_chat
omw
opinion_lexicon
paradigms
pe08
pil
pl196x
ppattach
problem_reports
product_reviews_1
product_reviews_2
pros_cons
ptb
qc
rte
senseval
sentence_polarity
sentiwordnet
shakespeare
sinica_treebank
smultron
state_union
stopwords
subjectivity
swadesh
switchboard
timit
toolbox
treebank
twitter_samples
udhr
udhr2
unicode_samples
verbnet
verbnet3
webtext
wordnet
wordnet_ic
words
ycoe
```

#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

### LINKED ISSUE #3369: Adds Stable Baselines Python LIbrary
#### Potential Solution
Adds [stable-baselines](https://stable-baselines.readthedocs.io/en/master/) library to supported Python libraries:
"_Stable Baselines is a set of improved implementations of Reinforcement Learning (RL) algorithms based on OpenAI Baselines._"
```
sudo apt-get update && sudo apt-get install cmake libopenmpi-dev python3-dev zlib1g-dev
pip install stable-baselines
```

#### Checklist
- [x] I have completely filled out this template
- [ ] I have confirmed that this issue exists on the current `master` branch
- [ ] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
- [ ] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by AlexCatarino**
#### Description
Whilelists Python Libraries:
- copulas, hmm and pomegranate (closes #2832)
- tsfresh (closes #3084)
- LightGBM (closes #3173)
- BeautifulSoup (closes #3178)
- Facebook/Prophet (closes #3199)
- fastai (closes #3233)
- Pyramid ARIMA (closes #3234)
- punkt for nltk (closes #3367)
- Stable Baselines (closes #3369)

- Updates:
  - Miniconda to 4.5.12 (Foundation and travis)
  - Python to 3.6.8
  - numpy to 1.16.1

- Use `apt-get install -y openjdk-8-jdk openjdk-8-jre` to install JRE
- Revert changes to Dockerfile

#### Related Issue
Incorporate and closes #3368 #3334

#### Motivation and Context
Fix docker foundation build and add python libraries 

#### Requires Documentation Change
No.

#### How Has This Been Tested?
- Build docker image successfully in Linux
- Unit tests were added to verify the package installation.

#### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [c] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Non-functional change (xml comments/documentation/etc)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,12 +5,15 @@ mono:
...
+  - wget https://cdn.quantconnect.com/miniconda/Miniconda3-4.5.12-Linux-x86_64.sh
+  - bash Miniconda3-4.5.12-Linux-x86_64.sh -b
+  - rm -rf Miniconda3-4.5.12-Linux-x86_64.sh
   - sudo ln -s $HOME/miniconda3/lib/libpython3.6m.so /usr/lib/libpython3.6m.so
   - conda update -y python conda pip
-  - conda install -y cython pandas scipy
+  - conda install -y python=3.6.8
+  - conda install -y numpy=1.16.1
+  - conda install -y pandas=0.23.4
+  - conda install -y cython scipy
```
Comment (jaredbroad) [inline .travis.yml]: Should we tag the versions of these to prevent changes?

File: Dockerfile [hunk 2]
```diff
@@ -17,7 +17,7 @@ MAINTAINER QuantConnect <contact@quantconnect.com>
 #	sed -i 's/4.5/4.0/' Algorithm.VisualBasic/QuantConnect.Algorithm.VisualBasic.vbproj && \
 #	wget https://nuget.org/nuget.exe && \
 #	mono nuget.exe restore QuantConnect.Lean.sln -NonInteractive && \
-#	xbuild /property:Configuration=Release && \
+#	msbuild /property:Configuration=Release && \
```
Comment (jaredbroad) [inline Dockerfile]: :+1:  Please also update the readme while changing this.

File: DockerfileLeanFoundation [hunk 3]
```diff
@@ -44,25 +40,40 @@ RUN echo "deb http://download.mono-project.com/repo/ubuntu stable-xenial/snapsho
...
+# Install supported third party python packages
+RUN conda update -y conda pip
+RUN conda install -y python=3.6.8
+RUN conda install -y msgpack-python=0.5.6
+RUN conda install -y numpy=1.16.1
+RUN conda install -y pandas=0.23.4
+RUN conda install -y wrapt=1.10.11
+RUN conda install -y dill=0.2.9
+RUN conda install -y jsonschema lxml
+RUN conda install -y blaze cython cvxopt keras nltk tensorflow
```
Comment (jaredbroad) [inline DockerfileLeanFoundation]: Suggest all of these use specifically requested versions or we'll pay for it later :D

File: DockerfileLeanFoundation [hunk 4]
```diff
@@ -44,25 +40,40 @@ RUN echo "deb http://download.mono-project.com/repo/ubuntu stable-xenial/snapsho
...
+RUN conda install -y -c pytorch pytorch torchvision
+RUN conda install -y -c pytorch -c fastai nvidia-ml-py3 fastai
```
Comment (jaredbroad) [inline DockerfileLeanFoundation]: pytorch repeated 3x?

File: Tests/Python/PythonPackagesTests.cs [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+        }
+
+        [Test]
+        public void FastAiTest()
+        {
+            AssetCode(
+                @"
+from fastai.text import *
+def RunTest():
+    return 'Test is only importing the module, since available tests take too long'"
```
Comment (jaredbroad) [inline Tests/Python/PythonPackagesTests.cs]: Hillarious "fastai" test axed because it takes too long.

File: DockerfileLeanFoundation (hunk 4 above)
Comment (AlexCatarino) [reply DockerfileLeanFoundation]: `conda install -c` is for [channel customization](https://docs.conda.io/projects/conda/en/latest/commands/install.html#Channel%20Customization).

File: Tests/Python/PythonPackagesTests.cs (hunk 5 above)
Comment (AlexCatarino) [reply Tests/Python/PythonPackagesTests.cs]: Hopefully, we will get a use-case for algorithm trading that can be added here later.

File: Dockerfile (hunk 2 above)
Comment (mchandschuh) [reply Dockerfile]: using `msbuild` now? I didn't know that was available for *nix systems.

File: Tests/Python/PythonPackagesTests.cs [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+*/
+
+using NUnit.Framework;
+using Python.Runtime;
+using System;
+
+namespace QuantConnect.Tests.Python
+{
+    [TestFixture, Category("TravisExclude")]
+    public class PythonPackagesTests
```
Comment (mchandschuh) [inline Tests/Python/PythonPackagesTests.cs]: 👏 

I love it! I  think where possible, we should favor unit tests over regression tests as they provide better documentation, run faster, and are more focused on the system under test. Great work @AlexCatarino!

**Pull request merged**